File: exchange-core/src/main/java/exchange/core2/core/processors/InternalTransferProcessor.java
package exchange.core2.core.processors;

import exchange.core2.core.common.CoreCurrencySpecification;
import exchange.core2.core.common.CoreSymbolSpecification;
import exchange.core2.core.common.MatcherEventType;
import exchange.core2.core.common.MatcherTradeEvent;
import exchange.core2.core.common.UserProfile;
import exchange.core2.core.common.cmd.CommandResultCode;
import exchange.core2.core.common.cmd.OrderCommand;
import exchange.core2.core.orderbook.OrderBookEventsHelper;

/**
 * INTERNAL_TRANSFER —— 用户间同币种原子转账（Branch B）。
 * 字段映射：uid=from / size=to / symbol=currency / price=amount / orderId=transactionId。
 *
 * <p>from 与 to 可跨 shard，三阶段（同 {@link ResetFeeCommandProcessor} 骨架）：
 * R1 {@link #collectInput}（from-shard 校验 + 扣款 + 幂等，置 VALID 放行）→ ME {@link #buildMatcherEvents}（产事件）
 * → R2 {@link #applyEvent}（to-shard 入账，未知 to 自动建 SUSPENDED 档）。
 *
 * <p>唯一会失败的 from-NSF 压在 R1、只依赖本地状态；扣款按 seq 定序，无双花、无需冻结。
 * 守恒中性（accounts 桶内 from−x/to+x，不碰 adjustments），幂等锚在 from。
 */
public final class InternalTransferProcessor extends TwoStepCommandProcessor {

    /** R1/R2 实例（RiskEngine 每 shard 一份）。 */
    public InternalTransferProcessor(RiskEngine riskEngine) {
        super(null, riskEngine);
    }

    /** ME-stage 实例（MatchingEngineRouter 持）。 */
    public InternalTransferProcessor(OrderBookEventsHelper eventsHelper) {
        super(eventsHelper, null);
    }

    /** R1：仅 from-shard 校验 + 扣款 + 幂等。 */
    @Override
    public void collectInput(OrderCommand cmd) {
        super.collectInput(cmd);
        if (!riskEngine.uidForThisHandler(cmd.uid)) {
            return;
        }
        final long toUid = cmd.size;
        final int currency = cmd.symbol;
        final long amount = cmd.price;
        if (cmd.uid == toUid) {
            cmd.resultCode = CommandResultCode.INTERNAL_TRANSFER_INVALID_SELF;
            return;
        }
        if (amount <= 0) {
            cmd.resultCode = CommandResultCode.RISK_INVALID_AMOUNT;
            return;
        }
        final UserProfile from = riskEngine.getUserProfileService().getUserProfile(cmd.uid);
        if (from == null) {
            cmd.resultCode = CommandResultCode.AUTH_INVALID_USER;
            return;
        }
        // NSF 与提现同口径；不 claim，可改额同 id 重试
        if (riskEngine.withdrawableBalance(from, currency) < amount) {
            cmd.resultCode = CommandResultCode.RISK_NSF;
            return;
        }
        if (!from.processedTransactionIds.tryClaim(cmd.orderId, cmd.timestamp)) {
            cmd.resultCode = CommandResultCode.USER_MGMT_ACCOUNT_BALANCE_ADJUSTMENT_ALREADY_APPLIED_SAME;
            return;
        }
        emitSnapshot(cmd, cmd.uid, from, currency, from.accounts.addToValue(currency, -amount));
        cmd.resultCode = CommandResultCode.VALID_FOR_MATCHING_ENGINE;
    }

    /** ME：把 to / currency / amount 打进一条事件。 */
    @Override
    protected void buildMatcherEvents(OrderCommand cmd) {
        final MatcherTradeEvent ev = eventsHelper.newMatcherEvent();
        ev.eventType = MatcherEventType.INTERNAL_TRANSFER_EVENT;
        ev.matchedOrderUid = cmd.size; // to
        ev.price = cmd.symbol;         // currency
        ev.size = cmd.price;           // amount
        ev.nextEvent = null;
        cmd.matcherEvent = ev;
    }

    /** R2：to-shard 入账（未知 to 自动建 SUSPENDED 档）。 */
    @Override
    public void applyEvent(OrderCommand cmd, MatcherTradeEvent ev, CoreSymbolSpecification spec,
        CoreCurrencySpecification currencySpec) {
        super.applyEvent(cmd, ev, spec, currencySpec);
        if (ev.eventType != MatcherEventType.INTERNAL_TRANSFER_EVENT) {
            return;
        }
        final long toUid = ev.matchedOrderUid;
        if (!riskEngine.uidForThisHandler(toUid)) {
            return;
        }
        final int currency = (int) ev.price;
        final UserProfile to = riskEngine.getUserProfileService().getUserProfileOrAddSuspended(toUid);
        emitSnapshot(cmd, toUid, to, currency, to.accounts.addToValue(currency, ev.size));
    }

    /** 发一条转账后余额快照事件（free = 新余额 − 冻结）。 */
    private void emitSnapshot(OrderCommand cmd, long uid, UserProfile up, int currency, long newBalance) {
        final long locked = riskEngine.calculateLocked(up, currency);
        riskEngine.getEventsHelper().sendInternalTransferEvent(cmd, uid, currency, newBalance - locked, locked);
    }
}

File: exchange-core-rs/src/core/processors/internal_transfer_processor.rs
//! 对应 Java `InternalTransferProcessor`（两步处理器）：`INTERNAL_TRANSFER` 用户间同币种原子转账。
//! 字段映射：`cmd.uid=fromUid`、`cmd.size=toUid`（overloaded）、`cmd.symbol=currency`、`cmd.price=amount`、`cmd.order_id=transactionId`。R1 校验+立即扣款，merge 1:1 直传，R2 入账（收款方不存在则建 SUSPENDED 档）；守恒 from-=amount/to+=amount。

use crate::core::common::cmd::command_result_code::CommandResultCode;
use crate::core::processors::risk_engine::RiskEngine;
use crate::core::processors::symbol_specification_provider::SymbolSpecificationProvider;
use crate::core::processors::user_profile_service::UserProfileService;

/// 无状态处理器——零大小类型，不持有字段。
pub struct InternalTransferProcessor;

impl InternalTransferProcessor {
    /// R1：对应 Java `collectInput`。校验顺序：self→amount<=0→from 缺失→NSF→幂等（try_claim_tx）；成功后立即 `from.accounts[currency] -= amount`。
    #[allow(clippy::too_many_arguments)]
    pub fn collect_input(
        engine: &RiskEngine,
        ups: &mut UserProfileService,
        ssp: &SymbolSpecificationProvider,
        from_uid: i64,
        to_uid: i64,
        currency: i32,
        amount: i64,
        order_id: i64,
        timestamp: i64,
    ) -> CommandResultCode {
        if from_uid == to_uid {
            return CommandResultCode::InternalTransferInvalidSelf;
        }
        if amount <= 0 {
            return CommandResultCode::RiskInvalidAmount;
        }
        let from = match ups.get_mut(from_uid) {
            Some(u) => u,
            None => return CommandResultCode::AuthInvalidUser,
        };
        if engine.withdrawable_balance(from, currency, ssp) < amount {
            return CommandResultCode::RiskNsf;
        }
        if !from.try_claim_tx(order_id, timestamp) {
            return CommandResultCode::UserMgmtAccountBalanceAdjustmentAlreadyAppliedSame;
        }
        from.add_to_account(currency, -amount);
        CommandResultCode::Success
    }

    /// merge：对应 Java `buildMatcherEvents`。1:1 直传，调用方已确认 R1 成功，不重复校验。
    pub fn build_matcher_events(to_uid: i64, currency: i32, amount: i64) -> (i64, i32, i64) {
        (to_uid, currency, amount)
    }

    /// R2：对应 Java `applyEvent`。`to` 从未见过则自动建 SUSPENDED 档，随后入账。
    pub fn apply_event(ups: &mut UserProfileService, to_uid: i64, currency: i32, amount: i64) {
        let to = ups.get_or_add_suspended(to_uid);
        to.add_to_account(currency, amount);
    }
}
